File: chapter1. 오브젝트와 의존관계/chapter1_tlswltjq/src/main/java/org/example/chapter1_tlswltjq/dao/UserDao.java
package org.example.chapter1_tlswltjq.dao;

import org.example.chapter1_tlswltjq.domain.User;

import java.sql.*;

public class UserDao {
    private ConnectionMaker connectionMaker; //구체클래스를 알 필요 엉ㅂㅅ다.

    public UserDao(ConnectionMaker connectionMaker) {
        this.connectionMaker = connectionMaker;
    }

    public void add(User user) throws ClassNotFoundException, SQLException {
        Connection c = connectionMaker.makeConnection();//이제 인터페이스에 정의 된 메소드를 사용하므로 클래스의 변경에 따른 메소드명 변화도 걱정할 필요없다.
        PreparedStatement ps = c.prepareStatement(
                "insert into users(id, name, password) values(?,?,?)"
        );
        ps.setString(1, user.getId());
        ps.setString(2, user.getName());
        ps.setString(3, user.getPassword());

        ps.executeUpdate();

        ps.close();
        c.close();
    }

    public User get(String id) throws SQLException, ClassNotFoundException {
        Connection c = connectionMaker.makeConnection();
        PreparedStatement ps = c.prepareStatement(
                "select * from users where id = ?"
        );

        ps.setString(1, id);

        ResultSet rs = ps.executeQuery();

        rs.next();
        User user = new User();
        user.setId(rs.getString("id"));
        user.setName(rs.getString("name"));
        user.setPassword(rs.getString("password"));

        rs.close();
        ps.close();
        c.close();

        return user;
    }
}

File: chapter1. 오브젝트와 의존관계/chapter1_albireo3754/src/main/kotlin/UserDao.kt
package com.albireo3754

import java.sql.Connection
import java.sql.DriverManager

class UserDao(private val connectionMaker: SimpleConnectionMaker) {
    fun add(user: User) {
        val connection = connectionMaker.getConnection();

        var ps = connection.prepareStatement("insert into users(id, name, password) values(?, ?, ?)")
        ps.setString(1, user.id);
        ps.setString(2, user.name);
        ps.setString(3, user.password);

        ps.executeUpdate()

        ps.close()
        connection.close()
    }

    fun get(id: String): User {
        val connection = connectionMaker.getConnection();

        var ps = connection.prepareStatement("select * from users where id = ?")
        ps.setString(1, id);

        val rs = ps.executeQuery()
        rs.next()

        val user = User(rs.getString("id"), rs.getString("name"), rs.getString("password"))

        rs.close()
        ps.close()
        connection.close()

        return user
    }


}